File: backend/src/ai/reward_engine.py
def calculate_reward(data):
    """
    Calculates the reward points based on the item type, condition, and trust score.

    Args:
        data (dict): A dictionary containing the item details.
            - item (str): The type of item.
            - condition (str): The condition of the item.
            - trustScore (int): The trust score of the user.

    Returns:
        dict: A dictionary containing the total reward points and the reasons.
            - coins (int): The total reward points.
            - reasons (list): A list of strings explaining the reward points.
    """
    item_points = {
        "Plastic Bottle": 20,
        "Glass Bottle": 25,
        "Metal Can": 30,
        "Paper": 10,
        "Battery": 50
    }

    item = data["item"]
    condition = data["condition"]
    trust_score = data["trustScore"]

    coins = item_points.get(item, 0)
    reasons = [f"{item}: {coins} points"]

    if condition == "Clean":
        coins += 5
        reasons.append("Clean condition bonus: 5 points")

    if trust_score >= 80:
        coins += 5
        reasons.append("High trust score bonus: 5 points")

    return {"coins": coins, "reasons": reasons}


if __name__ == "__main__":
    test_data = {
        "item": "Plastic Bottle",
        "condition": "Clean",
        "trustScore": 85
    }

    result = calculate_reward(test_data)
    print("\n========== RESULT ==========\n")

    for key, value in result.items():
        if isinstance(value, list):
            print(f"{key}:")
            for item in value:
                print(f"  • {item}")
        else:
            print(f"{key}: {value}")

    print("\n============================")

File: backend/src/ai/rewardEngine.js
/**
 * Calculates the reward points for a given item based on its condition and the user's trust score.
 *
 * @param {Object} data - The input data object.
 * @param {string} data.item - The type of item (e.g., "Plastic Bottle", "Glass Bottle", "Metal Can", "Paper", "Battery").
 * @param {string} data.condition - The condition of the item (e.g., "Clean", "Dirty").
 * @param {number} data.trustScore - The trust score of the user (0-100).
 *
 * @returns {Object} An object containing the calculated reward points and the reasons for the reward.
 * @returns {number} reward.coins - The total reward points earned.
 * @returns {string[]} reward.reasons - An array of strings describing the reasons for the reward.
 */
function calculateReward(data) {
    // Define the base reward points for each item type
    const itemPoints = {
      "Plastic Bottle": 20,
      "Glass Bottle": 25,
      "Metal Can": 30,
      "Paper": 10,
      "Battery": 50
    };
  
    // Extract the item, condition, and trustScore from the input data
    const { item, condition, trustScore } = data;
  
    // Initialize the reward object with the base points and reasons
    const reward = {
      coins: itemPoints[item] || 0, // Use the base points for the item, or 0 if not found
      reasons: [item] // Add the item type as the first reason for the reward
    };
  
    // Check if the item condition is "Clean" and add a bonus
    if (condition === "Clean") {
      reward.coins += 5; // Add 5 bonus points for a clean item
      reward.reasons.push("Clean Bonus"); // Add "Clean Bonus" as a reason for the reward
    }
  
    // Check if the user's trust score is greater than or equal to 80 and add a bonus
    if (trustScore >= 80) {
      reward.coins += 5; // Add 5 bonus points for a high trust score
      reward.reasons.push("High Trust Bonus"); // Add "High Trust Bonus" as a reason for the reward
    }
  
    // Return the final reward object
    return reward;
  }
  
  // Export the calculateReward function using CommonJS
  module.exports = calculateReward;